File: Exo5.4.py
import threading
import random
import sys

cancel_flag = False

class Tableau:
    def __init__(self, n):
        self.tab = [0]*n
        self.n = n

def generer_tableau(T):
    for i in range(T.n):
        T.tab[i] = random.randint(0, 99)
        if cancel_flag:
            return

def verifier_inclusion(T1, T2):
    included = all(x in T2.tab for x in T1.tab)
    print(f"T1 {'est inclus dans' if included else 'n\'est pas inclus dans'} T2")

def surveiller_entrée():
    global cancel_flag
    c = input("Appuyez sur 'A' pour annuler : ")
    if c.upper() == 'A':
        cancel_flag = True
        print("Programme annulé par l'utilisateur !")
        sys.exit(0)

# Initialisation
T1 = Tableau(5)
T2 = Tableau(10)

th_cancel = threading.Thread(target=surveiller_entrée)
th_cancel.start()

th1 = threading.Thread(target=generer_tableau, args=(T1,))
th2 = threading.Thread(target=generer_tableau, args=(T2,))

th1.start()
th2.start()

th1.join()
th2.join()

# Vérification inclusion
th_incl = threading.Thread(target=verifier_inclusion, args=(T1, T2))
th_incl.start()
th_incl.join()

File: Exo5.4.c
#include <stdio.h>
#include <stdlib.h>
#include <pthread.h>
#include <time.h>
#include <unistd.h>

typedef struct
{
    int* tab;
    int n;
} Tableau;

int cancel_flag = 0; // global pour interruption

// Génération de tableau aléatoire
void* generer_tableau(void* arg) 
{
    int i;
    Tableau* t = (Tableau*) arg;
    t->tab = malloc(t->n * sizeof(int));
    if (!t->tab) 
    {
        return (NULL);
    }    
    for (i = 0; i < t->n; i++) 
    {
        t->tab[i] = rand() % 100;
        if (cancel_flag)
        {
            return (NULL); // interruption
        }    
    }
    return (NULL);
}

// Vérification si T1 est inclus dans T2
void* verifier_inclusion(void* arg) 
{
    Tableau* tab = (Tableau*) arg;
    Tableau* T1 = &tab[0];
    Tableau* T2 = &tab[1];
    int i,j;
    int included = 1;
    for (i = 0; i < T1->n; i++) 
    {
        int found = 0;
        for (j = 0; j < T2->n; j++) 
        {
            if (T1->tab[i] == T2->tab[j]) 
            {
                found = 1;
                break;
            }
        }
        if (!found) 
        {
            included = 0;
            break;
        }
        if (cancel_flag) 
        {
            return (NULL); // interruption
    }
    printf("T1 %s T2\n", included ? "est inclus dans" : "n'est pas inclus dans");
    return (NULL);
}

// Thread pour surveiller entrée utilisateur
void* surveiller_entrée(void* arg) 
{
    char c;
    printf("Appuyez sur 'A' pour annuler...\n");
    c = getchar();
    if (c == 'A' || c == 'a') 
    {
        cancel_flag = 1;
        printf("Programme annulé par l'utilisateur !\n");
        exit(0);
    }
    return (NULL);
}

int main() 
{
    srand(time(NULL));

    Tableau T1, T2;
    T1.n = 5;  // plus petit
    T2.n = 10; // plus grand

    pthread_t th1, th2, th_incl, th_cancel;

    pthread_create(&th_cancel, NULL, surveiller_entrée, NULL);
    pthread_create(&th1, NULL, generer_tableau, &T1);
    pthread_create(&th2, NULL, generer_tableau, &T2);

    pthread_join(th1, NULL);
    pthread_join(th2, NULL);

    // Vérifier inclusion
    Tableau tab_arr[2] = {T1, T2};
    pthread_create(&th_incl, NULL, verifier_inclusion, tab_arr);
    pthread_join(th_incl, NULL);

    free(T1.tab);
    free(T2.tab);

    return (0);
}